### PR TITLE
Feature: encryption with context

### PR DESCRIPTION
**Pull request by robholland**
## What was changed:

The encryption example is updated to take key configuration from the workflow/activity context. A propagator is added so that the activities see the encryption context set for the workflow.

Notes:

This encryption is not applied to anything other than workflow/activity inputs/results. Specifically headers, memo, search attributes and errors are not encrypted.

The web UI does not understand the encoding used so will only show the first few characters of the crypt text for the inputs/results. This means operators are not able to copy/paste the payloads to decrypt them locally.

tctl does not understand the encoding used so will not show the inputs/results at all when showing the workflow history.

## Why?

Some encryption use cases require varying keys based on some aspect of the workflow.

## How has this been tested?

Tested manually by running the sample workflow and checking the history in the Web UI/tctl.

### CODE REVIEW

File: cryptconverter/plugin/main.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	"github.com/hashicorp/go-plugin"
+	"github.com/temporalio/samples-go/cryptconverter"
+	"go.temporal.io/sdk/converter"
+	"go.temporal.io/server/tools/cli"
+)
+
+var handshakeConfig = plugin.HandshakeConfig{
+	ProtocolVersion:  1,
+	MagicCookieKey:   "TEMPORAL_PLUGIN_DATA_CONVERTER",
+	MagicCookieValue: "cookie",
```
Comment (Sushisource) [inline cryptconverter/plugin/main.go:13]: I think it's probably best practice to make this a little more application-specific, even though this is a sample people are probably going to copy-paste things like this wholesale

File: cryptconverter/data-converter.go [hunk 2]
```diff
@@ -1,94 +1,95 @@
...
-	if err != nil {
-		return nil, err
-	}
-
-	gcm, err := cipher.NewGCM(c)
-	if err != nil {
-		return nil, err
-	}
+func (dc *CryptDataConverter) getKey(keyId string) (key []byte) {
+	// Key can be fetched from KMS or other secure storage.
```
Comment (Sushisource) [inline cryptconverter/data-converter.go:66]: nit: I'd probably make this wording stronger. Maybe even "*must* be fetched ... in a production setting"

**Review by Sushisource — APPROVED** [review]
Looks great, thanks Rob! Just a few minor wording changes

File: cryptconverter/cryptconverter.go [hunk 3]
```diff
@@ -36,5 +36,12 @@ func Workflow(ctx workflow.Context, name string) (string, error) {
 func Activity(ctx context.Context, name string) (string, error) {
 	logger := activity.GetLogger(ctx)
 	logger.Info("Activity", "name", name)
+
+	activity.RecordHeartbeat(ctx, "Thinking...")
+	// Uncomment if you'd like to check the heartbeat is encrypted in UI/tctl.
+	// Heartbeat details aren't saved to the history so this state won't be available
+	// once the activity completes.
```
Comment (mfateev) [inline cryptconverter/cryptconverter.go:43]: You can test this differently. Let activity timeout and the result of the last heartbeat will be included in the timeout error.

File: go.mod [hunk 4]
```diff
@@ -2,22 +2,23 @@ module github.com/temporalio/samples-go
 
 go 1.16
 
+replace go.temporal.io/sdk => /Users/rob/Code/github.com/robholland/sdk-go
```
Comment (feedmeapples) [inline go.mod:5]: leftover

**Review by feedmeapples — CHANGES_REQUESTED** [review]

File: go.mod (hunk 4 above)
Comment (feedmeapples) [inline go.mod:5]:
```suggestion
```

File: go.mod [hunk 5]
```diff
@@ -2,22 +2,23 @@ module github.com/temporalio/samples-go
...
+
+replace go.temporal.io/server => /Users/rob/Code/github.com/robholland/temporal
```
Comment (feedmeapples) [inline go.mod:7]:
```suggestion
```

File: go.mod [hunk 6]
```diff
@@ -2,22 +2,23 @@ module github.com/temporalio/samples-go
 
 go 1.16
 
+replace go.temporal.io/sdk => /Users/rob/Code/github.com/robholland/sdk-go
+
```
Comment (feedmeapples) [inline go.mod:6]:
```suggestion
```

File: go.mod [hunk 7]
```diff
@@ -2,22 +2,23 @@ module github.com/temporalio/samples-go
...
+replace go.temporal.io/server => /Users/rob/Code/github.com/robholland/temporal
+
```
Comment (feedmeapples) [inline go.mod:8]:
```suggestion
```

File: go.mod [hunk 8]
```diff
@@ -2,22 +2,21 @@ module github.com/temporalio/samples-go
 
 go 1.16
 
+replace go.temporal.io/server => github.com/robholland/temporal v1.1.1-0.20210427090742-bc9685c643ea
```
Comment (feedmeapples) [inline go.mod:5]: this needs to be cleaned before merging ^

File: go.mod [hunk 9]
```diff
@@ -2,22 +2,21 @@ module github.com/temporalio/samples-go
...
+	github.com/stretchr/testify v1.7.0
 	github.com/uber-go/tally v3.3.17+incompatible
 	github.com/uber/jaeger-client-go v2.25.0+incompatible
 	github.com/uber/jaeger-lib v2.4.0+incompatible // indirect
-	go.temporal.io/api v1.4.0
-	go.temporal.io/sdk v1.5.0
-	gopkg.in/yaml.v3 v3.0.0-20210106172901-c476de37821d
+	go.temporal.io/api v1.4.1-0.20210420220407-6f00f7f98373
+	go.temporal.io/sdk v1.6.0
+	go.temporal.io/server v0.0.0-00010101000000-000000000000
```
Comment (feedmeapples) [inline go.mod:20]: and this ^

**Review by feedmeapples — CHANGES_REQUESTED** [review]

File: cryptconverter/README.md [hunk 10]
```diff
@@ -1,14 +1,20 @@
 ### Steps to run this sample:
 1) You need a Temporal service running. See details in README.md
-2) Run the following command to start the worker
+2) Compile the cryptconverter plugin for tctl
+```
+go build -o ./cryptconverter-plugin plugin/main.go
```
Comment (feedmeapples) [inline cryptconverter/README.md:5]: let's add the resulting binary to gitignore as the built artifact

**Review by feedmeapples — APPROVED** [review]

**Pull request merged**